File: onyx-database-tests/test/embedded/diskmap/MultiThreadTest.java
package embedded.diskmap;

import category.EmbeddedDatabaseTests;
import com.onyx.diskmap.DefaultMapBuilder;
import com.onyx.diskmap.MapBuilder;
import org.junit.Test;
import org.junit.experimental.categories.Category;

import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Created by timothy.osborn on 3/27/15.
 */
@Category({ EmbeddedDatabaseTests.class })
public class MultiThreadTest extends AbstractTest {

    ExecutorService pool = Executors.newFixedThreadPool(9);

    static Random randomGenerator = new Random();
    @Test
    public void testMultiThread()
    {
        MapBuilder store = new DefaultMapBuilder(TEST_DATABASE);
        Map<Integer, Object> myMap = store.getHashMap("second");

        long time = System.currentTimeMillis();
        List<Future> items = new ArrayList<>();
        for(int i = 1; i < 20; i++)
        {
            final int p = randomGenerator.nextInt((19 - 1) + 1) + 1;

            if((i % 5) == 0)
            {
                Runnable scanThread = new Runnable() {
                    @Override
                    public void run()
                    {
                        Iterator it = myMap.entrySet().iterator();

                        while (it.hasNext())
                        {
                            Map.Entry entry = (Map.Entry) it.next();
                            entry.getKey();
                        }
                    }
                };

                items.add(pool.submit(scanThread));
            }
            MyRunnable runnable = new MyRunnable(myMap, p);
            items.add(pool.submit(runnable));
        }

        items.stream().forEach(
                future ->
                {
                    try
                    {
                        future.get();
                    } catch (InterruptedException e)
                    {
                        e.printStackTrace();
                    } catch (ExecutionException e)
                    {
                        e.printStackTrace();
                    }
                }
        );

        System.out.println("Done in " + (time - System.currentTimeMillis()));
        store.close();
    }

    class MyRunnable implements Runnable
    {
        protected Map myMap = null;
        int p;

        public MyRunnable(Map myMap, int p)
        {
            this.myMap = myMap;
            this.p = p;

        }

        @Override
        public void run()
        {
            final int it = p;
            for(int k = 1; k < 5000; k++)
            {
                Integer val = k * p;
                myMap.put(val, "HIYA" + val);
                String strVal = (String)myMap.get(val);
                myMap.remove(val);
                strVal = (String)myMap.get(val);
            }
        }
    }

}

File: onyx-database-tests/test/diskmap/MultiThreadTest.kt
package diskmap

import com.onyx.diskmap.factory.impl.DefaultDiskMapFactory
import database.base.DatabaseBaseTest
import org.junit.BeforeClass
import org.junit.Test

import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.Future
import kotlin.test.assertNull

/**
 * Created by timothy.osborn on 3/27/15.
 */
class MultiThreadTest {


    companion object {
        private var threadPool = Executors.newFixedThreadPool(10)

        private val TEST_DATABASE = "C:/Sandbox/Onyx/Tests/multiThreadMapTest.db"

        @BeforeClass
        @JvmStatic
        fun beforeTest() = DatabaseBaseTest.deleteDatabase(TEST_DATABASE)

    }

    @Test
    fun testMultiThread() {
        val store = DefaultDiskMapFactory(TEST_DATABASE)
        val myMap = store.getHashMap<MutableMap<Int, String>>(Int::class.java, "second")

        val items = ArrayList<Future<*>>()
        for (i in 1..19) {
            val p = DatabaseBaseTest.randomInteger + 1

            if (i % 5 == 0) {
                val scanThread = DatabaseBaseTest.async(threadPool) {
                    val it = myMap.entries.iterator()

                    while (it.hasNext()) {
                        val entry = it.next() as Map.Entry<Int, String>
                        entry.key
                    }
                }
                items.add(scanThread)
            }

            val runnable = MyRunnable(myMap, p)
            items.add(DatabaseBaseTest.async(threadPool, runnable::run))
        }

        items.forEach { it.get() }
        store.close()
    }

    internal inner class MyRunnable(private var myMap: MutableMap<Int, String>, private var p: Int) : Runnable {

        override fun run() {
            for (k in 1..4999) {
                val value:Int = (k * p)
                myMap.put(value, "HIYA" + value)
                myMap[value] as String
                myMap.remove(value)
                assertNull(myMap[value])
            }
        }
    }
}
